--- old/postgresql-setup-8.7/bin/postgresql-setup.in	2022-01-04 09:22:07.000000000 +0000
+++ new/postgresql-setup-8.7/bin/postgresql-setup.in	2023-01-30 11:25:40.671231953 +0000
@@ -167,7 +167,7 @@
     test -w "$initdb_log" || echo "$initdb_log is not writeable by $USER"
 
     # Initialize the database
-    initdbcmd+=( "$PGENGINE"/initdb --pgdata="$pgdata" --auth=ident )
+    initdbcmd+=( "$PGENGINE"/initdb --pgdata="$pgdata" --auth-host=md5 )
     eval "initdbcmd+=( $PGSETUP_INITDB_OPTIONS )"
     "${initdbcmd[@]}" >> "$initdb_log" 2>&1 < /dev/null
 
@@ -192,6 +192,58 @@
     test -f "$pgdata/PG_VERSION"
 }
 
+initbabelfish()
+{
+    info $"Initializing Babelfish extensions"
+
+    # Preload TDS library
+    sed -i "s|^[[:space:]#]*shared_preload_libraries[[:space:]]=[^#]*|shared_preload_libraries = 'babelfishpg_tds' |g" "$pgdata/postgresql.conf"
+
+    # Enable SSL for MS ODBC Driver 18
+    # https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/891
+    # https://www.postgresql.org/docs/14/ssl-tcp.html
+    pushd $pgdata >/dev/null
+    openssl req -new -x509 -days 3650 -nodes -text -out server.crt -keyout server.key -subj "/CN=localhost" >/dev/null 2>&1
+    popd >/dev/null
+    if [ -f "$pgdata/server.crt" ] && [ -f "$pgdata/server.key" ] ; then
+        sed -i "s|^[[:space:]#]*ssl[[:space:]]=[^#]*|ssl = on |g" "$pgdata/postgresql.conf"
+    else
+        warn "Error generating SSL certificate, SSL is disabled"
+    fi
+
+    # Create and init Babelfish extension, avoid single-user mode
+    # https://www.postgresql.org/message-id/87zj93agba.fsf%40news-spur.riddles.org.uk
+    @bindir@/pg_ctl start -D $pgdata
+    if test $? -ne 0; then
+        error "Babelfish initialization error: pg_ctl failure"
+        return 1
+    fi
+    @bindir@/psql -U "$USER" -d postgres -a << EOF
+CREATE USER babelfish WITH SUPERUSER CREATEDB CREATEROLE PASSWORD 'babelfish' INHERIT;
+CREATE DATABASE babelfish OWNER babelfish;
+\c babelfish
+SET allow_system_table_mods = ON;
+CREATE EXTENSION IF NOT EXISTS "babelfishpg_tds" CASCADE;
+GRANT ALL ON SCHEMA sys to babelfish;
+ALTER SYSTEM SET babelfishpg_tsql.database_name = 'babelfish';
+ALTER DATABASE babelfish SET babelfishpg_tsql.migration_mode = 'multi-db';
+SELECT pg_reload_conf();
+CALL sys.initialize_babelfish('babelfish');
+EOF
+    if test $? -ne 0; then
+        error "Babelfish initialization error: psql failure"
+        @bindir@/pg_ctl stop -D $pgdata
+        return 1
+    fi
+    @bindir@/pg_ctl stop -D $pgdata
+
+    info $"Babelfish initialization complete, use 'systemctl start postgresql' to start the server"
+    info $"ODBC connection:"
+    info $"DRIVER={ODBC Driver 18 for SQL Server};SERVER=127.0.0.1,1433;UID=babelfish;PWD=babelfish;TrustServerCertificate=yes"
+    info $"JDBC connection with mssql-jdbc:"
+    info $"jdbc:sqlserver://127.0.0.1:1433;trustServerCertificate=true"
+    warn $"DB superuser 'babelfish' was created with password 'babelfish', please change the password before allowing remote connections."
+}
 
 initdb()
 {
@@ -206,6 +258,7 @@
         error $"Initializing database failed, possibly see $initdb_log"
         script_result=1
     fi
+    initbabelfish
 }
 
 
