--- ./postgresql_modified_for_babelfish-BABEL_2_2_0__PG_14_5/postgresql-setup-8.7/bin/postgresql-setup.in	2022-01-04 09:22:07.000000000 +0000
+++ ./postgresql_modified_for_babelfish-BABEL_2_2_0__PG_14_5/postgresql-setup-8.7/bin/postgresql-setup.in	2022-12-26 18:50:02.688314995 +0000
@@ -167,7 +167,7 @@
     test -w "$initdb_log" || echo "$initdb_log is not writeable by $USER"
 
     # Initialize the database
-    initdbcmd+=( "$PGENGINE"/initdb --pgdata="$pgdata" --auth=ident )
+    initdbcmd+=( "$PGENGINE"/initdb --pgdata="$pgdata" --auth-host=md5 )
     eval "initdbcmd+=( $PGSETUP_INITDB_OPTIONS )"
     "${initdbcmd[@]}" >> "$initdb_log" 2>&1 < /dev/null
 
@@ -189,6 +189,21 @@
         fi
     fi
 
+    # Preload TDS library
+    sed -i "s|^[[:space:]#]*shared_preload_libraries[[:space:]]=[^#]*|shared_preload_libraries = 'babelfishpg_tds' |g" "$pgdata/postgresql.conf"
+
+    # Enable SSL for MS ODBC Driver 18
+    # https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/891
+    # https://www.postgresql.org/docs/14/ssl-tcp.html
+    pushd $pgdata >/dev/null
+    openssl req -new -x509 -days 3650 -nodes -text -out server.crt -keyout server.key -subj "/CN=localhost" >/dev/null 2>&1
+    popd >/dev/null
+    if [ -f "$pgdata/server.crt" ] && [ -f "$pgdata/server.key" ] ; then
+        sed -i "s|^[[:space:]#]*ssl[[:space:]]=[^#]*|ssl = on |g" "$pgdata/postgresql.conf"
+    else
+        echo "Error generating SSL certificate, SSL is disabled"
+    fi
+
     test -f "$pgdata/PG_VERSION"
 }
 
